source/common/stats/deferred_creation.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats.h"
+
+#include "source/common/common/cleanup.h"
+#include "source/common/common/thread.h"
+#include "source/common/stats/symbol_table.h"
+#include "source/common/stats/utility.h"
+
+namespace Envoy {
+namespace Stats {
+
+/**
+ * Lazy-initialization wrapper for StatsStructType, intended for deferred instantiation of a block
+ * of stats that might not be needed in a given Envoy process.
+ *
+ * This class is thread-safe -- instantiations can occur on multiple concurrent threads.
+ * This is used when
+ * :ref:`enable_deferred_creation_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_deferred_creation_stats>` is enabled.
+ */
+template <typename StatsStructType>
+class DeferredCreation : public DeferredCreationCompatibleInterface<StatsStructType> {
+public:
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  DeferredCreation(const typename StatsStructType::StatNameType& stat_names,
+                   Stats::ScopeSharedPtr scope)
+      : initialized_([&scope]() -> Gauge& {
```

Comment:
do we need to capture scope by value here?

That this works makes me suspect we don't have a test that removes the scope first and then instantiates the stats.

Reply:
The constructor accepts the scope by value, so this should be fine, right?

Reply:
the constructor will be out of scope before the lambda gets run, so no help there :(

I think maybe this is all moot because StatsStructType does not hold a reference-count for the scope, but holds it by `Scope&`. This is because I only (relatively) recently changed scopes to be represented as shared_ptr but didn't change all the `Scope&` everywhere in Envoy to a `ScopeSharedPtr`.

So what will happen in the test scenario I just thought of:
```
ScopeSharedPtr scope = createScope(...);
DeferredCreation deferred_stats(stat_names, scope);
scope.reset();
deferred_stats.instantiate();
```

I think the code as is will reference freed memory, but with my suggestion it will work, but not be useful :)

It will successfully instantiate the stats in the scope, and then the Cleanup thing will remove the scope. So you'll be left with struct full of invalid stats. So if you then do:

```
deferred_stats->my_counter_.inc():
````
that will then crash.

This is beyond the scope of this PR to resolve (sic) but I'd still vote for capturing scope by value on line 32.

Reply:
The scope is used first in initialize the gauge, and then moved into the ctor_, could you help me to understand the life cycle issue there? 

the scope is always supposed to outlive the stats, otherwise the stats will be deleted IIUC.

Reply:
I think I articulated the lifecycle issue above.

In practice it doesn't matter but my recommendation stands -- just change `&scope` to `scope` in your capture.

Reply:
I am not following where this "reference freed memory" is from. 
the line 32 capture is used and thrown away after init of the initialized_ Gauge.

Scope is held in ctor_ until it's called. 
the case raised by @jmarantz should have no problem to call instantiate(), because the scope was captured. 

but further references to its member I am not sure, since the last copy of "scope" is deleted. 

I have the understanding that scope always outlive the individual StatStruct, if that not true, then the previous suggestion of do not capture scope in DeferredCreation need to be revisited.

Reply:
The lambda in line 32 is not stored anywhere, so capturing scope by value there does not change anything. It's a temporary that is invoked and immediately destroyed.

The lambda that is stored is the one on line 38, and the scope there is (effectively) captured by value (moved from the passed in by value object for the constructor of `DeferredCreation`.
My point though is that too is destroyed when `instantiate` is called as `ctor_` is set to `nullptr` when invoked.

However storing the scope in `DeferredCreation` would also require storing it in `DirectStats` or else we could see unexpected behaviour in the future based on the value of `enableDeferredCreationStats`

Reply:
With the code as is, the call to `deferred_stats.instantiate()` will work just fine, but `deferred_stats->my_counter_.inc()` will crash (but this is the same behaviour for direct_stats).

Reply:
added comment as asked.